Fix explore library and change rl_clientlib to use explore lib to swap actions

rajan-chari opened the pull request:
1) sample_after_normalizing will always normalize pdf
2) remove sample_after_normalizing that takes scores
3) add swap_chosen() to swap chosen action with first action
4) added unit tests
5) refactored ranking_response.  added a const iterator.
6) rl_clientlib will use swap from explore library to switch action ordering

ataymano commented on explore/explore_internal.h:
```diff
@@ -294,56 +298,31 @@ namespace exploration
+  int swap_chosen(ActionIt action_first, ActionIt action_last, std::forward_iterator_tag action_category, uint32_t chosen_index)
   {
-    if (pdf_last < pdf_first || ranking_last < ranking_first)
+    if ( action_last < action_first )
```
[nit] Looks like all E_EXPLORATION_BAD_RANGE checks (304-312) can be collapsed into: if (action_first + chosen_index >= action_last)

rajan-chari replied:
Had to revert this change since not all iterators have all comparison operators.  Moved back to conservative implementation.

ataymano commented on reinforcement_learning/include/ranking_response.h:
```diff
@@ -129,31 +132,77 @@ namespace reinforcement_learning {
+    std::string _event_id;
+    size_t _chosen_action_id;
+    std::string _model_id;
+    using coll_type = std::vector<action_prob>;
```
[nit] I think we have "_t" everywhere else for type naming

rajan-chari replied:
Done.

---
In reply to: [228962208](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1658#discussion_r228962208) [](ancestors = 228962208)

ataymano commented on reinforcement_learning/rlclientlib/vw_model/vw_model.cc:
```diff
@@ -43,19 +43,24 @@ namespace reinforcement_learning { namespace model_management {
       // Pick a slot using the pdf. NOTE: sample_after_normalizing() can change the pdf
       uint32_t chosen_index;
-      auto const scode = e::sample_after_normalizing(rnd_seed, std::begin(pdf), std::end(pdf), chosen_index);
+      auto scode = e::sample_after_normalizing(rnd_seed, std::begin(pdf), std::end(pdf), chosen_index);
```
the whole code block {action_ids, pdf} -> ranking_response looks copy-pasted in vw_model and live_model_impl. Isn't it better to dedup it in some helper class / method?

rajan-chari replied:
There is a subtle diff.  one case uses indexes, the other uses action_id[idx].  I found a bug here so thanks for the heads up.

jackgerrits commented on explore/explore.h:
```diff
@@ -89,42 +89,14 @@ namespace exploration {
-  int sample_after_normalizing(uint64_t seed, PdfIt pdf_first, PdfIt pdf_last, InputScoreIt scores_first, InputScoreIt scores_last, OutputIt ranking_first, OutputIt ranking_last);
+  * @brief Swap the first value with the chosen index.  
+  *
+  * @tparam ActionIt Iterator type of the action. Must be an forward_iterator.
```
>an [](start = 58, length = 2)

nit: a

rajan-chari replied:
Fixed

---
In reply to: [228987853](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1658#discussion_r228987853) [](ancestors = 228987853)

jackgerrits commented on explore/explore_internal.h:
```diff
@@ -257,19 +257,23 @@ namespace exploration
-    chosen_index = i - 1;
+    if(!index_found)
+      chosen_index = i - 1;
+    
```
nit: whitespace

rajan-chari replied:
fixed

---
In reply to: [228988454](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1658#discussion_r228988454) [](ancestors = 228988454)

jackgerrits commented on explore/explore_internal.h:
```diff
@@ -294,56 +298,31 @@ namespace exploration
-      InputScoreIt scores_first, InputScoreIt scores_last, std::random_access_iterator_tag scores_category,
-      OutputIt ranking_first, OutputIt ranking_last, std::random_access_iterator_tag ranking_category)
+  template<typename ActionIt>
+  int swap_chosen(ActionIt action_first, ActionIt action_last, std::forward_iterator_tag action_category, uint32_t chosen_index)
```
>action_category [](start = 89, length = 15)

This variable is ununsed?

JohnLangford commented:
This looks good to me.  Ready to merge?

rajan-chari commented:
@JohnLangford I was making changes related to comments.  Also found an issue, updated rl sample app and added more unit tests.  Should be good to go once the build finishes.

JohnLangford commented:
Merged, thank you :-)
